server/util/bytebufferpool/bytebufferpool.go

```diff
@@ -5,23 +5,48 @@ import (
 	"sync"
 )
 
+// safeSyncPool is a wrapper around `sync.Pool` that provides type safety
+// using Go generics.
+type safeSyncPool[T any] struct {
```

Comment:
This feels a little overkill for internal usage, but it's fine.
We could move this somewhere else in the repo and use it everywhere we use sync.Pool, then it would feel less wasteful. Of course we don't need to do that in this PR.